src/libexpr/flake/flake.cc

```diff
@@ -441,7 +452,16 @@ LockedFlake lockFlake(
                     throw Error("cannot update flake input '%s' in pure mode", inputPathS);
 
                 if (input.isFlake) {
-                    auto inputFlake = getFlake(state, *input.ref, lockFlags.useRegistries, flakeCache);
+
+                    /* If this input is due to a user --update-input
+                       request, ensure that caching is disabled.
+                       Normally URL fetches are cached for a period of
+                       time (see 'nix show-config | grep tarball-ttl')
+                       but if this was an explicit user request for an
+                       update, ignore any recently cached fetch. */
+                    auto inputFlake = lockFlags.inputUpdates.count(inputPath)
```

Comment:
I'm not sure I understand your question: without this conditional the entire patch is irrelevant.  This is the main place where the functionality of this PR is actually introduced: this checks that `--update-input` was specified for the corresponding `inputPath` and if so, ensures that the fetch cache is bypassed so that the input is actually retrieved from the remote rather than the cache.

The `--recreate-lock-file` has slightly different semantics as I understand it based on the existing flake code, and it does not necessarily imply an `--update-input` for each input.  To change those semantics would need logic to be implemented elsewhere and presumably affect the population `inputUpdates` contents at that point, but that's beyond the scope of the intended effect of  this PR and should probably be handled separately.

Reply:
Sorry if that was unclear. My understanding of the code is that we can reach this point either when passing `--update-input` or becase there's no previous lockfile (or `--recreate-lock-file` is passed which I think has the same effect). This PR only disables the cache in the first case, but maybe we want to disable it in every case. After all, this cache is mostly here so that you can do `nix build -f channel:nixos-unstable hello` without re-downloading the `nixos-unstable` channel each time, and it might not make sense at all with flakes.

Reply:
The interaction between `--recreate-lock-file` (or the lack of a lock file) and the `--update-input` argument is ill-defined and the subject of some dispute (c.f. issue number 3779).  I'd prefer to keep this PR specifically targeted and not get caught up in unresolved design decisions.